scripts/combineCards.py

```diff
@@ -1,122 +1,226 @@
-nuisanceEdits = [];
+obsline = []
+obskeyline = []
+keyline = []
+expline = []
+systlines = {}
+signals = []
+backgrounds = []
+shapeLines = []
+paramSysts = {}
+flatParamNuisances = {}
+discreteNuisances = {}
+groups = {}
+rateParamsPerCard = []
+extArgs = {}
+binParFlags = {}
+bpf_new2old = {}
+nuisanceEdits = []
 constraint_terms = []
 
-def compareParamSystLines(a,b):
-  if float(a[0])!=float(b[0]): return False
-  if "/" in a[1]:
-    if "/" not in b[1] : return False
-    a1,a2 = a[1].split("/")
-    b1,b2 = b[1].split("/")
-    if float(a1)!=float(b1): return False
-    if float(a2)!=float(b2): return False
-  else:
-    if float(a[1])!=float(a[1]): return False
-  return True
+
+def compareParamSystLines(a, b):
+    if float(a[0]) != float(b[0]):
+        return False
+    if "/" in a[1]:
+        if "/" not in b[1]:
+            return False
+        a1, a2 = a[1].split("/")
+        b1, b2 = b[1].split("/")
+        if float(a1) != float(b1):
+            return False
+        if float(a2) != float(b2):
+            return False
+    else:
+        if float(a[1]) != float(a[1]):
+            return False
+    return True
 
 
-cmax = 5 # column width
+cmax = 5  # column width
 if not args:
-    raise RuntimeError, "No input datacards specified."
-for ich,fname in enumerate(args):
-    label = "ch%d" % (ich+1)
-    if "=" in fname: (label,fname) = fname.split("=")
-    fname = options.fprefix+fname
+    raise RuntimeError("No input datacards specified.")
+for ich, fname in enumerate(args):
+    label = "ch%d" % (ich + 1)
+    if "=" in fname:
+        (label, fname) = fname.split("=")
+    fname = options.fprefix + fname
     dirname = os.path.dirname(fname)
     if fname.endswith(".gz"):
         import gzip
+
         file = gzip.open(fname, "rb")
         fname = fname[:-3]
     else:
         file = open(fname, "r")
     DC = parseCard(file, options)
-    singlebin = (len(DC.bins) == 1)
+    singlebin = len(DC.bins) == 1
     if label == ".":
-        label=DC.bins[0] if singlebin else "";
+        label = DC.bins[0] if singlebin else ""
     elif not singlebin:
-        label += "_";
+        label += "_"
     for b in DC.bins:
-        bout = label if singlebin else label+b
-        b_in  = label if singlebin else b
-        if isVetoed(b_in,options.channelVetos): continue
-        if not isIncluded(b_in,options.channelIncludes): continue
+        bout = label if singlebin else label + b
+        b_in = label if singlebin else b
+        if isVetoed(b_in, options.channelVetos):
+            continue
+        if not isIncluded(b_in, options.channelIncludes):
+            continue
         obskeyline.append(bout)
-        for (p,e) in DC.exp[b].items(): # so that we get only self.DC.processes contributing to this bin
-            if DC.isSignal[p] == False: continue
-            #print "in DC.exp.items:b,p", b,p
+        for (p, e) in DC.exp[b].items():  # so that we get only self.DC.processes contributing to this bin
```

Comment:
Due to https://portingguide.readthedocs.io/en/latest/dicts.html#changed-key-order the ordering of the processes within a bin for combined cards will change with the migration. In general it should remain stable for py3.7+ but the language does not guarantee any particular dict key order. The fundamental issue is that `DC.exp[b]` is a dict, and if deterministic card combination is desired then this should be a list of tuples instead.

Reply:
I agree, but we should keep this in mind for the future; some optimization for running stuff relies and are more effective depending on the ordering.

Reply:
can't it just be an OrderedDict to retain dict-like access while really being a list of tuples under the hood?

Reply:
It could be, but I guess my point here is that no matter what we cannot deterministically recreate the previous behavior.